### PR TITLE
Change the display method of Popup Overlay on Windows

### LINKED ISSUE #1478: [BUG] Popup CanBeDismissedByTappingOutsideOfPopup = false not modal and doesn't prevent from interacting with Parent Page
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Did you read the "Reporting a bug" section on Contributing file?

- [X] I have read the "Reporting a bug" section on Contributing file: https://github.com/CommunityToolkit/Maui/blob/main/CONTRIBUTING.md#reporting-a-bug

### Current Behavior

If I call a popup where I set the CanBeDismissedByTappingOutsideOfPopup = false; the result is this and it's not modal, I can interact with the page.
image
But if I set the parameters as true, the page gets a layer that prevents the interaction and graphically sets a difference between the popup and the application.
image

### Expected Behavior

Have the same behavior as between the two value of the parameter with the difference to not being able to be closed on tapping outside the popup.

### Steps To Reproduce

1. Create a MAUI project with the creation of a maui community toolkit popup. 
2. Call a Popup with CanBeDismissedByTappingOutsideOfPopup = false; an one with CanBeDismissedByTappingOutsideOfPopup = true;

### Link to public reproduction project repository

[Link to Repository](https://github.com/FedePella94/BUG-Popup-MAUI-Community-Toolkit)

### Environment

```markdown
- .NET MAUI CommunityToolkit: 6.0
- OS: Windows 10.0.19041.0
- .NET MAUI: 7.0
```

### Anything else?

_No response_

### PR DESCRIPTION
**Pull request by cat0363**
This PR solves the following problems.
1. An issue where the NavigationBar area is operable when the Popup's CanBeDismissedByTappingOutsideOfPopup property is False.
2. An issue where overlays are not displayed on the Popup background when the Popup's CanBeDismissedByTappingOutsideOfPopup property is False.

 ### Description of Change ###

If CanBeDismissedByTappingOutsideOfPopup is False, the ContentPage area is blocked from interaction, but the NavigationBar area is not blocked from interaction. 
By specifying LightDismissOverlayMode.On for the LightDismissOverlayMode property, it is possible to display an overlay that covers the NavigationBar area, but it can be closed with a click. 
In the previous implementation using Flyout, this was prevented using the OnClosing event, but this is not possible with Popup because the OnClosing event cannot be captured.

To solve this problem, I implement our own overlay. 
I created a PopupOverlay class that inherits from the WindowOverlay class.

[src\CommunityToolkit.Maui.Core\Views\Popup\PopupOverlay.windows.cs]

    public class PopupOverlay : WindowOverlay
    {
        IWindowOverlayElement popupOverlayElement;

        public PopupOverlay(IWindow window, Color? overlayColor = null) : base(window)
        {
            popupOverlayElement = new PopupOverlayElement(this, overlayColor);
            AddWindowElement(popupOverlayElement);
            EnableDrawableTouchHandling = true;
        }

        class PopupOverlayElement : IWindowOverlayElement
        {
            readonly IWindowOverlay overlay;
            RectF overlayRect = new RectF();
            Color overlayColor = Color.FromRgba(255, 255, 255, 153);

            public PopupOverlayElement(IWindowOverlay overlay, Color? overlayColor = null)
            {
                this.overlay = overlay;
                if (overlayColor != null)
                {
                    this.overlayColor = overlayColor;
                }
            }

            public bool Contains(Point point)
            {
                return this.overlayRect.Contains(new Point(point.X / overlay.Density, point.Y / overlay.Density));
            }

            public void Draw(ICanvas canvas, RectF dirtyRect)
            {
                this.overlayRect = dirtyRect;
                canvas.FillColor = overlayColor;
                canvas.FillRectangle(dirtyRect.X, dirtyRect.Y, dirtyRect.Width, dirtyRect.Height);
            }
        }
    }

Below are the changes when opening and closing Popup.

[src\CommunityToolkit.Maui.Core\Handlers\Popup\PopUpHandler.windows.cs]

    public static void MapOnOpened(PopupHandler handler, IPopup view, object? result)
    {
        ArgumentNullException.ThrowIfNull(view.Parent);
        ArgumentNullException.ThrowIfNull(handler.MauiContext);
        var parent = view.Parent.ToPlatform(handler.MauiContext);
        parent.IsHitTestVisible = false;
        handler.PlatformView.XamlRoot = parent.XamlRoot;
        handler.PlatformView.IsHitTestVisible = true;
        handler.PlatformView.IsOpen = true;

        var uiSetting = new UISettings();
        Windows.UI.Color backgroundColor = uiSetting.GetColorValue(UIColorType.Background);
        var window = view.GetWindow();
        window.AddOverlay(new PopupOverlay(window, IsColorDark(backgroundColor) ? Color.FromRgba(0, 0, 0, 153) : Color.FromRgba(255, 255, 255, 153)));

        view.OnOpened();
    }

    public static void MapOnClosed(PopupHandler handler, IPopup view, object? result)
    {
        var window = view.GetWindow();
        window.RemoveOverlay(window.Overlays.First());

        view.HandlerCompleteTCS.TrySetResult();
        handler.DisconnectHandler(handler.PlatformView);
    }

    public static void MapCanBeDismissedByTappingOutsideOfPopup(PopupHandler handler, IPopup view)
    {
        handler.PlatformView.IsLightDismissEnabled = view.CanBeDismissedByTappingOutsideOfPopup;
        handler.PlatformView.LightDismissOverlayMode = LightDismissOverlayMode.Off;
    }

    static bool IsColorDark(Windows.UI.Color clr)
    {
        return (((5 * clr.G) + (2 * clr.R) + clr.B) < (8 * 128));
    }

It is determined whether or not it is in dark mode when the popup is opened.
This determines whether the popup's overlay color is white or black.
By adding PopupOverlay to Window's Overlay, I am displaying my own overlay.
To display your own overlay, the LightDismissOverlayMode property must be turned off at all times.

    var uiSetting = new UISettings();
    Windows.UI.Color backgroundColor = uiSetting.GetColorValue(UIColorType.Background);
    var window = view.GetWindow();
    window.AddOverlay(new PopupOverlay(window, IsColorDark(backgroundColor) ? Color.FromRgba(0, 0, 0, 153) : Color.FromRgba(255, 255, 255, 153)));

When you close the Popup, the PopupOverlay is removed from the window's overlay.

    var window = view.GetWindow();
    window.RemoveOverlay(window.Overlays.First());

 

 ### Linked Issues ###
 

 - Fixes #1478

 ### PR Checklist ###
 
 - [x] Has a linked Issue, and the Issue has been `approved`(bug) or `Championed` (feature/proposal)
 - [ ] Has tests (if omitted, state reason in description)
 - [x] Has samples (if omitted, state reason in description)
 - [x] Rebased on top of `main` at time of PR
 - [x] Changes adhere to [coding standard](https://github.com/CommunityToolkit/Maui/blob/main/CONTRIBUTING.md#contributing-code---best-practices)
 - [ ] Documentation created or updated: https://github.com/MicrosoftDocs/CommunityToolkit/pulls 

 ### Additional information ###

Below are the verification results.
In validation, we are clicking on the NavigationBar's MenuBarItem and Popup background.

[Not dark mode]

https://github.com/CommunityToolkit/Maui/assets/125236133/1baa85f9-d847-4324-a21e-dc04044de70a

[Dark mode]

https://github.com/CommunityToolkit/Maui/assets/125236133/1d0444b2-3107-4b64-94c6-da1f704e4942

If the Popup's CanBeDismissedByTappingOutsideOfPopup property is False, you will notice that the NavigationBar area is not clickable and the Overlay is displayed on the Popup's background.

### CODE REVIEW

**Comment by cat0363** [comment]
The overlay colors are white and black with an Alpha value of 40%. 
Please make corrections as necessary.

**Comment by cat0363** [comment]
Overlays do not work correctly when popups are nested. . .

https://github.com/CommunityToolkit/Maui/assets/125236133/d95dd555-71ed-455b-90b4-77fa276f48f4

Overlay is placed below all Popups.

**Comment by cat0363** [comment]
When popups are nested, there is a problem that the overlay is not applied to cover the lower Popup, so the lower Popup can be manipulated. Even in the Popup before the modification, if CanBeDismissedByTappingOutsideOfPopup is False, lower Popups can be manipulated, so this may not be a problem due to the modification.

**Comment by cat0363** [comment]
If I check the description of the LightDismissOverlayMode property, it is a property for specifying whether to darken the area outside the Popup when the IsLightDismissEnabled property is True. Therefore, if the IsLightDismissEnabled property is False, the area outside the Popup will not be darkened. Maybe it's Popup's specifications...

File: src/CommunityToolkit.Maui.Core/Handlers/Popup/PopUpHandler.windows.cs [hunk 1]
```diff
@@ -146,4 +156,9 @@ void OnSizeChanged(object? sender, WindowSizeChangedEventArgs e)
 			PopupExtensions.SetLayout(PlatformView, VirtualView, MauiContext);
 		}
 	}
+
+	static bool IsColorDark(Windows.UI.Color clr)
```
Comment (VladislavAntonyuk) [inline src/CommunityToolkit.Maui.Core/Handlers/Popup/PopUpHandler.windows.cs:160]: do we have a similar method in ColorsConverter? can we reuse it?

File: src/CommunityToolkit.Maui.Core/Handlers/Popup/PopUpHandler.windows.cs [hunk 2]
```diff
@@ -36,6 +40,12 @@ public static void MapOnOpened(PopupHandler handler, IPopup view, object? result
 		handler.PlatformView.XamlRoot = parent.XamlRoot;
 		handler.PlatformView.IsHitTestVisible = true;
 		handler.PlatformView.IsOpen = true;
+
+		var uiSetting = new UISettings();
```
Comment (VladislavAntonyuk) [inline src/CommunityToolkit.Maui.Core/Handlers/Popup/PopUpHandler.windows.cs:44]: probably extract to the separate method AddOverlay

File: src/CommunityToolkit.Maui.Core/Handlers/Popup/PopUpHandler.windows.cs [hunk 3]
```diff
@@ -17,6 +18,9 @@ public partial class PopupHandler : ElementHandler<IPopup, Popup>
 	/// <param name="result">The result that should return from this Popup.</param>
 	public static void MapOnClosed(PopupHandler handler, IPopup view, object? result)
 	{
+		var window = view.GetWindow();
+		window.RemoveOverlay(window.Overlays.First());
```
Comment (VladislavAntonyuk) [inline src/CommunityToolkit.Maui.Core/Handlers/Popup/PopUpHandler.windows.cs:22]: Are we sure we always have at least one overlay?

File: src/CommunityToolkit.Maui.Core/Views/Popup/PopupOverlay.windows.cs [hunk 4]
```diff
@@ -0,0 +1,55 @@
...
+	class PopupOverlayElement : IWindowOverlayElement
+	{
+		readonly IWindowOverlay overlay;
+		RectF overlayRect = new RectF();
+		Color overlayColor = Color.FromRgba(255, 255, 255, 153);
+
+		public PopupOverlayElement(IWindowOverlay overlay, Color? overlayColor = null)
+		{
+			this.overlay = overlay;
+			if (overlayColor != null)
```
Comment (VladislavAntonyuk) [inline src/CommunityToolkit.Maui.Core/Views/Popup/PopupOverlay.windows.cs:37]:
```suggestion
			if (overlayColor is not null)
```

File: src/CommunityToolkit.Maui.Core/Views/Popup/PopupOverlay.windows.cs [hunk 5]
```diff
@@ -0,0 +1,55 @@
...
+		popupOverlayElement = new PopupOverlayElement(this, overlayColor);
+		AddWindowElement(popupOverlayElement);
+		EnableDrawableTouchHandling = true;
+	}
+
+	class PopupOverlayElement : IWindowOverlayElement
+	{
+		readonly IWindowOverlay overlay;
+		RectF overlayRect = new RectF();
+		Color overlayColor = Color.FromRgba(255, 255, 255, 153);
```
Comment (VladislavAntonyuk) [inline src/CommunityToolkit.Maui.Core/Views/Popup/PopupOverlay.windows.cs:32]: probably add comment it is 60% background

**Review by VladislavAntonyuk — COMMENTED** [review]
Thank you @cat0363 , I left some tiny comments

File: src/CommunityToolkit.Maui.Core/Handlers/Popup/PopUpHandler.windows.cs [hunk 6]
```diff
@@ -129,6 +140,13 @@ protected override void ConnectHandler(Popup platformView)
 		}
 		base.ConnectHandler(platformView);
 	}
+	
+	static void AddOverlayToWindow(IWindow window)
+	{
+		var uiSetting = new UISettings();
+		Windows.UI.Color backgroundColor = uiSetting.GetColorValue(UIColorType.Background);
+		window.AddOverlay(new PopupOverlay(window, IsColorDark(backgroundColor) ? Color.FromRgba(0, 0, 0, 153) : Color.FromRgba(255, 255, 255, 153)));
```
Comment (brminnick) [inline src/CommunityToolkit.Maui.Core/Handlers/Popup/PopUpHandler.windows.cs:148]:
```suggestion
		window.AddOverlay(new PopupOverlay(window, IsColorDark(backgroundColor) ? Color.FromRgba(0, 0, 0, 153) : Color.FromRgba(255, 255, 255, 153))); // 60% Opacticy
```

File: src/CommunityToolkit.Maui.Core/Views/Popup/PopupOverlay.windows.cs (hunk 5 above)
Comment (brminnick) [reply src/CommunityToolkit.Maui.Core/Views/Popup/PopupOverlay.windows.cs:32]:
```suggestion
		Color overlayColor = Color.FromRgba(255, 255, 255, 153); // 60% Opacity 
```

**Review by brminnick — APPROVED** [review]
Thanks @cat0363 + @VladislavAntonyuk!

**Pull request merged**